### ISSUE TITLE
-d does not return back the original series

### ISSUE THREAD

**Issue by aigarspetresevics** [fracdiff/fracdiff#127]
Using the sklearn api, the differenced series does not go back to its original by using the negative order of differencing with which it was differenced in the first place.

Please let us know how to undifference.

Thanks

**Comment by ghost** [comment]
That's a great question. while mathematically fracdiff with negative order gives rise to an inverse, i.e. (1-D)^d (1-D)^(-d) = 1 where D is a lag operator, computationally it would hold only approximately due to the truncation at the finite window.
also, in general, differentiation can lose information--e.g. in d=1 case, integration of a differentiated time series has an ambiguity of integration constant.

https://github.com/simaki/fracdiff/issues/78

**Comment by aigarspetresevics** [comment]
HI simaki,

Thanks for getting back to me. It does not hold even approximately, the inverse is very very far of... After trying to implement this idea I'm starting to get confused about the motivation.  We use this to improve our models performance, achieving stationarity without sacrificing memory. But the idea is to convert back to the original scale after doing prediction. Now say we used 10k+ timesteps to compute our differentiated series, then trained our model then predicted 20 days ahead. It would be impossible to convert the prediction results back to the original scale since the window would be different, and the computational method for 10k days would be different from 20 days.

Could you please shed some light on how to convert back to the original scale after prediction? I suppose one thing it can give us is the general trend regardless of scale.

**Comment by ghost** [comment]
I personally don't take the slogan that fracdiff keeps the memory of time series at the face value. fracdiff does lose some memory in exchange for taming nonstationarity - the advantage is just that this exchange is relatively less than diff.

the approximation is quite rough - you would need a quite wide window to keep, say, 90% of the memory.
https://github.com/fracdiff/fracdiff/blob/main/fracdiff/sklearn/tol.py#L56-L59
also, fracdiff erases the additive scale of time series - fracdiffs of X and X+c are the same. 

thus it wouldn't always be feasible to "convert back to the original scale after prediction". i would use fracdiff as just a feature/signal

**Comment by aigarspetresevics** [comment]
Yes that makes sense and thanks for getting back to me. I run some tests and using differentiated prices as feature does slightly improve the results. I’ll try run diff on some other non stationary features and see how my model behaves then.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by ghost** [comment]
glad if it helped. do not hesitate to reopen this if you get something to discuss further.

**Comment by microprediction** [comment]
Hmmm, interesting topic. It seems to me that one should be able to cook up a better approximate inverse by matching coefficients etc, rather than assuming that -d is the closest thing to an inverse, which is clearly isn't after truncation.

**Comment by ghost** [comment]
that is a good point. the formulation should depend on what 'close' means but one way is to solve an inverse problem D^d Y = X, where X is the vector of the original series, D is a lag operator, and Y = "D^-d X" is what we want to evaluate, which boils down to an inverse of a convolution matrix. this leads to yet another formulation of the fractional differentiation of time series, where fractional differentiation with d>=0 is defined as before whereas d<0 is by such an inverse. for an infinite window limit the formulation is the same as before while for finite window it may worth trying to look what difference does this make.

**Comment by microprediction** [comment]
I have some motivation to get this to work. I've added a reminder to [myself](https://github.com/microprediction/timemachines/issues/96) and maybe I can make a small contrib.
